mono/metadata/threads.c

```diff
@@ -234,154 +235,110 @@ mono_thread_get_abort_prot_block_count (MonoInternalThread *thread)
-			} else {
-				new_state |= INTERRUPT_REQUEST_DEFERRED_BIT;
-			}
-		}
+	/* pending exception, self abort */
+	if (state & INTERRUPT_SYNC_REQUESTED_BIT)
+		return TRUE;
 
-		//bounds check abort_prot_count
-		g_assert (new_val >= 0);
-		g_assert (new_val < (1 << ABORT_PROT_BLOCK_BITS));
-		new_state |= new_val << ABORT_PROT_BLOCK_SHIFT;
+	/* abort, interruption, suspend */
+	if ((state & INTERRUPT_ASYNC_REQUESTED_BIT) && !(state & ABORT_PROT_BLOCK_MASK))
+		return TRUE;
 
-	} while (InterlockedCompareExchangePointer ((volatile gpointer)&thread->thread_state, (gpointer)new_state, (gpointer)old_state) != (gpointer)old_state);
-	return (new_state & INTERRUPT_REQUESTED_BIT) == INTERRUPT_REQUESTED_BIT;
+	return FALSE;
 }
 
-
-//Don't use this function, use inc/dec below
-static void
-mono_thread_abort_prot_block_count_add (MonoInternalThread *thread, int val)
+gboolean
+mono_threads_end_abort_protected_block (void)
 {
+	MonoInternalThread *thread = mono_thread_internal_current ();
 	gsize old_state, new_state;
 	do {
 		old_state = thread->thread_state;
-		verify_thread_state (old_state);
 
-		int new_val = val + ((old_state & ABORT_PROT_BLOCK_MASK) >> ABORT_PROT_BLOCK_SHIFT);
 		//bounds check abort_prot_count
+		int new_val = ((old_state & ABORT_PROT_BLOCK_MASK) >> ABORT_PROT_BLOCK_SHIFT) - 1;
 		g_assert (new_val >= 0);
 		g_assert (new_val < (1 << ABORT_PROT_BLOCK_BITS));
-		new_state = (old_state & ~ABORT_PROT_BLOCK_MASK) | (new_val << ABORT_PROT_BLOCK_SHIFT);
 
+		new_state = old_state - (1 << ABORT_PROT_BLOCK_SHIFT);
 	} while (InterlockedCompareExchangePointer ((volatile gpointer)&thread->thread_state, (gpointer)new_state, (gpointer)old_state) != (gpointer)old_state);
-}
 
-static void
-mono_thread_inc_abort_prot_block_count (MonoInternalThread *thread)
-{
-	mono_thread_abort_prot_block_count_add (thread, 1);
-}
-
-static void
-mono_thread_dec_abort_prot_block_count (MonoInternalThread *thread)
-{
-	mono_thread_abort_prot_block_count_add (thread, -1);
+	return mono_thread_state_has_interruption (new_state);
 }
 
 static gboolean
 mono_thread_get_interruption_requested (MonoInternalThread *thread)
 {
 	gsize state = thread->thread_state;
-	return (state & INTERRUPT_REQUESTED_BIT) == INTERRUPT_REQUESTED_BIT;
+
+	return mono_thread_state_has_interruption (state);
 }
 
-/* Returns TRUE is there was a state change */
+/*
+ * Returns TRUE is there was a state change
+ * We clear a single interruption request, sync has priority.
+ */
 static gboolean
 mono_thread_clear_interruption_requested (MonoInternalThread *thread)
 {
 	gsize old_state, new_state;
 	do {
 		old_state = thread->thread_state;
-		verify_thread_state (old_state);
 
 		//Already cleared
-		if (!(old_state & (INTERRUPT_REQUESTED_BIT | INTERRUPT_REQUEST_DEFERRED_BIT)))
+		if (!(old_state & INTERRUPT_REQUESTED_MASK))
 			return FALSE;
-		new_state = old_state & ~(INTERRUPT_REQUESTED_BIT | INTERRUPT_REQUEST_DEFERRED_BIT);
+
+		if (old_state & INTERRUPT_SYNC_REQUESTED_BIT)
```

Comment:
If both bits are set then also thread_interruption_requested would have been bumped twice and the exeptions will be handled one after the other. First the sync interruption (also as it can be seen in mono_thread_execute_interruption, pending exceptions have priority) then the async interruption

The function is meant to clear only one interruption (which the code will process afterwards). I forgot to check for the abort count for the async case